EventGenerator/src/RPCGun_module.cc

```diff
@@ -94,176 +101,140 @@ namespace mu2e {
 
         _hmomentum     = tfdir.make<TH1F>( "hmomentum", "Produced photon momentum", 100,  40.,  140.  );
-
-        if(generateInternalConversion_){
+        if(RPCType_ == "mu2eInternalRPC"){
           _hElecMom  = tfdir.make<TH1F>("hElecMom" , "Produced electron momentum", 140,  0. , 140.);
           _hPosiMom  = tfdir.make<TH1F>("hPosiMom" , "Produced positron momentum", 140,  0. , 140.);
           _hMee      = tfdir.make<TH1F>("hMee"     , "M(e+e-) "           , 200,0.,200.);
           _hMeeVsE   = tfdir.make<TH2F>("hMeeVsE"  , "M(e+e-) vs E"       , 200,0.,200.,200,0,200);
           _hMeeOverE = tfdir.make<TH1F>("hMeeOverE", "M(e+e-)/E "         , 200, 0.,1);
-          _hy        = tfdir.make<TH1F>("hy"       , "y = (ee-ep)/|pe+pp|", 200,-1.,1.);
+          _hy        = tfdir.make<TH1F>("hy"       , "y = (ee-ep)/|pe+pp|", 200,-1.,1.);      
         }
       }
-
-    }
-
-  //================================================================
-
-  void RPCGun::beginRun(art::Run& run) {
-    protonPulse_.reset( new ProtonPulseRandPDF( eng_, protonPset_ ) );
   }
-
-  void RPCGun::produce(art::Event& event) {
-
-    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
-    std::unique_ptr<FixedTimeMap> timemap(new FixedTimeMap);
-
-    ConditionsHandle<AcceleratorParams> accPar("ignored");
-    double _mbtime = accPar->deBuncherPeriod;
-
-    IO::StoppedParticleTauNormF stop;
-    if (tmin_ > 0){
-      while (true){
-        const auto& tstop = stops_.fire();
-        timemap->SetTime(protonPulse_->fire());
-        if (tstop.t+timemap->time() < 0 || tstop.t+timemap->time() > tmin_){
-          if (applySurvivalProbability_){
-            double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-            if (weight > 1)
-              std::cout << "WEIGHT TOO HIGH " << weight << " " << fmod(tstop.t,_mbtime) << std::endl;
-            double rand = randomFlat_.fire();
-            if (weight > rand){
-              stop = tstop;
-              break;
-            }
-          }else{
-            stop = tstop;
-            break;
-          }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
+  
+  double RPCGun::MakeEventWeight(art::Ptr<SimParticle> part){ 
+      const PhysicsParams& gc = *GlobalConstantsHandle<PhysicsParams>();
+      double weight = 0.;
+      double tau = part->endProperTime() / gc.getParticleLifetime(part->pdgId());
+     
+      while(part->parent().isNonnull()) { //while particle has a parent which is not null
+        part = part->parent();
+        if ( abs(part->pdgId()) == PDGCode::pi_plus ) { //if pion
```

Comment:
Is it possible pions could have been allowed to decay in the simulation? For example, antiproton induced RPC doesn't turn off pion decays, so maybe there should be a flag for whether or not pion decays were turned off to produce this weight (unless in this scenario we just allow the weight to be produced but know it shouldn't be applied in the analysis)

Reply:
I don't know if there's a flag we can check in the G4 output if the pions were allowed to decay, I suggest you (Sophie) contact Krzysztof to see if there's a way to test that.  If so, the event weighting should be set to 1 in that case.

Reply:
I dont think there is but I can ask

Reply:
Hi Sophie,
I am not sure if I understand your question, but trying to answer the one I think you are asking,
as far as I know, the fact that the decay of a particle was affected,
either by changing the lifetime or removing the decay process, is not caried over between job stages.
I cc Andrei, in case he knows of an object which may preserve that information.
Krzysztof

Reply:
So it looks like G4 wont know. So we should state this somewhere and maybe have an on/off flag as a parameter - does that seem OK?

Reply:
I'm happy with an on/off switch, so the pbar RPC jobs for example would have their own fcl correctly configured for the different RPC case, but it's of course up to Dave and you. Thank you for looking into this!